ESPS/general/src/zero_pole/zero_pole.1
.\" Copyright (c) 1987, 1989, 1990 Entropic Speech, Inc.; All rights reserved
.\" @(#)zero_pole.1	3.6 5/24/91 ESI
.TH ZERO_POLE 1\-ESPS 5/24/91
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH NAME
zero_pole \- convert complex ascii zero and pole locations to ESPS filter 
.SH SYNOPSIS
.B zero_pole
[
.BI \-x " debug_level"
]
[
.BI \-c " comment"
]
[
.BI \-g " gain"
]
[
.BI \-f " freq"
]
.I zero_file
[
.I pole_file
]
.I feafilt_file
.SH DESCRIPTION
.PP
The program
.I zero_pole
reads complex zero locations from the file
.I zero_file
and (optionally) complex pole locations from the file
.I pole_file,
computes the numerator and denominator coefficients,
and prints the coefficients, poles and zeros to the output FEAFILT
file 
.I feafilt_file. 
Zero and pole locations are specified by a real part followed by an imaginary
part.
Poles or zeros on the real axis need the imaginary part specified.
If a pole or zero is specified off the real axis (i.e. with a nonzero
imaginary part), its complex conjugate is automatically included.
This insures that the resulting filter coefficients are all real.
Any poles specified outside of the unit circle will result in
an error.
.PP
Both input files are in ascii format.
The 
.I zero_file
and the
.I pole_file
may be the same file, in which case the numerator coefficients should appear first.
In both
.I zero_file
and
.I pole_file
the first number should be an integer which tells how many zeros or poles are in the
file.  The zeros or poles follow.  Both real and imaginary part of each pole and
zero must be specified.  Any combination of spaces and newlines may separate the
data.
.PP
The program prompts the user for comments to be added to the comment field of
.I feafilt_file
unless the \-c option is used or unless any coefficients are being read from the
standard input.  The user should type in his comments and terminate
them with a ^D on a line by itself.  In either case the user added comment is limited to 2047
characters.  In addition to any user specified comments, the program automatically
specifies the source of the zero locations and (if appropriate) the pole locations
in the comment field.
.PP
If
.I feafilt_file
is "\-", the output goes to the standard output.  If
.I zero_file
is "\-", the zeros are read from standard input.  If
.I pole_file
is "\-", the poles are read from standard input also.
.SH OPTIONS
The following options are supported:
.TP
.BI \-x " debug_level"
A value of 0 (the default value) will cause
.I zero_pole
to do its work silently, unless there is an error.
A nonzero value will cause various parameters to be printed out
during program initialization.
.br
.TP
.BI \-c " comment"
The character string
.I comment
is added to the comment field of the output file header.  If this option is not used,
the program prompts the user for a comment string, which he should terminate with
a ^D.  
.br
.TP
.BI \-g " gain"
This is the gain of the filter at the "critical frequency" specified
by the
.BI \-f
option.  The default for
.I gain
is 1.0.
If neither the 
.BI \-g 
nor the 
.BI \-f 
option is used, the filter gain is not normalized.
.br
.TP
.BI \-f " freq"
This is the critical frequency at which the gain of the filter is normalized, expressed as a fraction of the sampling frequency.
The default for
.I freq
is 0.0.
If neither the 
.BI \-g 
nor the
.BI \-f 
option is used, the filter gain is not normalized.
.SH ESPS PARAMETERS
The parameter file is not accessed.
.SH ESPS COMMON
The Common file is not processed.
.SH ESPS HEADER
A new header is created for the output FEAFILT file.
The program fills in appropriate values in the common part of the header.
The generic header items 
.I max_num
and
.I max_denom
are set to the number of items in 
.I num_file
and 
.I denom_file,
respectively;
.I define_pz 
is set to YES
and
.I complex_filter
is set to NO; 
.I type 
is set to FILT_ARB and
.I method
is set to PZ_PLACE.  All other items take their null values.
The program enters the names of the input coefficient files in the comment field of
the header.
.SH DIAGNOSTICS
The program prints a warning to the standard error output if the total comment
field of the output file exceeds MAX_STRING.
.SH SEE ALSO
atofilt(1\-ESPS), notch_filt(1\-ESPS), 
wmse_filt(1\-ESPS), iir_filt(1\-ESPS), fea2filt(1\-ESPS),
filter (1-ESPS), fft_filter (1-ESPS), 
pz_to_co (3-ESPSsp),
init_feafilt_hd(3\-ESPSu),
allo_feafilt_rec(3\-ESPSu), 
put_feafilt_rec(3\-ESPSu),
ESPS(5\-ESPS), FEA(5\-ESPS), FEAFILT(5\-ESPS), 
FILT (5-ESPS)
.SH AUTHOR
Brian Sublett. Modified for ESPS 3.0 by David Burton. Modified for
FEAFILT output by Bill Byrne.



ESPS/general/src/zero_pole/zero_pole.c
/*********************************************************
*
  This material contains proprietary software of Entropic Speech, Inc.   
  Any reproduction, distribution, or publication without the prior	   
  written permission of Entropic Speech, Inc. is strictly prohibited.
  Any public distribution of copies of this work authorized in writing by
  Entropic Speech, Inc. must bear the notice			
 								
    "Copyright (c) 1987, 1989, 1990 Entropic Speech, Inc.; All rights reserved"
*
*  Module Name: zero_pole.c
*
*  Written By:   Brian Sublett   1-30-87
*  Modified for ESPS 3.0 by David Burton
*  Modified to write FEAFILT files by Bill Byrne, 5/24/91
*  Checked by: 
*
*  Description: This program reads data from one or two ascii
*	        data files representing zeros and poles in the
*		complex plane.  It the computes the filter
*		coefficients and prints the results to a FILT
*		file.
*
*  
*
*********************************************************/

static char *sccs_id = "@(#)zero_pole.c	3.8 1/6/93  ESI";
char *Version="3.8", *Date="1/6/93";

/*
 * System Includes
*/

# include <stdio.h>
# include <math.h>

/*
 * ESPS Includes
*/
# include <esps/esps.h>
# include <esps/fea.h>
# include <esps/feafilt.h>
# include <esps/unix.h>
# include <esps/constants.h>
/*
 * Defines
*/
# define MX_S 2047  /* Number of characters in user entered comments. */
# define COM_S 50  /* Extra characters in default comments. */
# define SYNTAX	USAGE ("zero_pole [-x debug_level] [-c commnet] [-g gain] [-f freq] num_file [den_file] feafilt_file");
# define SMALL 1.0e-8

int debug_level =0;

main (argc, argv)
int argc;
char *argv[];
    {
    
    FILE *fpnum = stdin, *fpden, *fpout;    /* FILE stream ptrs -
					numerator, denominator, and output*/
    char *num_file = "<stdin>";		/* numerator coeff file name*/
    char *den_file;			/* denominator coeff filename*/
    char *filt_file;			/* output FILT filename*/
    struct header *oh;			/* FILT file header ptr*/
    struct feafilt *frec;		/* FEAFILT data structure ptr*/
    filtdesparams *fdp;
    int i, c;
    int nzero, npole = 0;		/* number of zeros and poles*/
    short cflag = NO;			/* Comment flag*/
    short dflag = NO;			/* denominator flag*/

    int maxnum, maxdenom;
    float *zptr, *pptr;
    short n_dim, d_dim;

    int  gain_flag = 0;  /* option flags*/
    double *za, *zb, *pa, *pb;		/* zero and pole locations*/
    double g = 1.0, f = 0.0, hypot (), sin (), cos ();
    double rpart, ipart, gain;
    char *comment1, *comment2;
    char   comment3[MX_S + 2];


    extern char *optarg;
    extern optind;


/* Check the command line options. */

    while ((c = getopt (argc, argv, "x:c:g:f:")) != EOF)
	{
	switch (c)
	    {
	    case 'x':
		if (sscanf(optarg,"%d", &debug_level) != 1)
		    {
		    Fprintf(stderr, "Error reading -x arg. Exitting.\n");
		    exit (1);
		    }
		if (debug_level != 0 && debug_level != 1)
		    {
		    debug_level = 1;
		    }
		break;
	    case 'c':
		if ((i = strlen (optarg)) <= MX_S)
		    Strcpy (comment3, optarg);
		else
		    {
		    Fprintf (stderr,"zero_pole: Comment field has %d characters\n", i);
		    Fprintf (stderr,"zero_pole: Maximum exceptable is %d\n", MX_S);
		    exit (1);
		    }
		if (comment3[i-1] != '\n')
		    {
		    comment3[i] = '\n';
		    comment3[i+1] = '\0';
		    }
		cflag = YES;
		break;
	    case 'g':
		if (sscanf (optarg, "%lf", &g) != 1)
		    {
		    Fprintf (stderr, "zero_pole: Error reading -g arg. Exitting.\n");
		    exit (1);
		    }
		gain_flag = 1;
		break;
	    case 'f':
		if (sscanf (optarg, "%lf", &f) != 1)
		    {
		    Fprintf (stderr, "zero_pole: Error reading -f arg. Exitting.\n");
		    exit (1);
		    }
		if (f < 0 || f > 1.0)
		    {
		    Fprintf (stderr,"zero_pole:f = %g invalid. f set to 0.\n", f);
		    }
		gain_flag = 1;
		break;
	    default:
		USAGE ("zero_pole [-x debug_level] feafilt_file num_file [den_file]");
	    }
	}

/* Get the filenames. */

    if(argc - optind < 2 || argc - optind > 3){
      SYNTAX;
      exit(1);
    }

 /*
  * Now get input numerator file name
 */

	{
	num_file = argv [optind++];
	if (strcmp (num_file, "-") == 0) 
	    {
	    num_file = "<stdin>";
	    cflag = YES; /*cannot get comment from stdin*/
	    }
	else 
	    {
	    TRYOPEN("zero_pole", num_file, "r", fpnum);
	    }
	}

    if (debug_level) Fprintf (stderr,
		"zero_pole: Numerator file is %s\n", num_file);

/* Read the number of zeros. */

    if (fscanf (fpnum,"%d", &nzero) != 1)
        {
        Fprintf (stderr,"zero_pole: Number of zeros not specified.\n");
	exit (1);
	}
    if (nzero < 0)
	{
	Fprintf (stderr,"nzero = %d is unacceptable - must be > zero.\n", 
				    nzero);
	exit (1);
	}
    if (debug_level)
        Fprintf (stderr,"nzero = %d\n", nzero);

/* Allocate space for the coefficients. */

    za = (double*) calloc ((unsigned)nzero, sizeof (double));
    spsassert(za != NULL, 
		"Couldn't allocate memory for zero coefficients\n");
    zb = (double*) calloc ((unsigned)nzero, sizeof (double));
    spsassert(zb != NULL, 
		"Couldn't allocate memory for zero coefficients\n");

/* read the zeros. */

    for (i=0; i<nzero; i++)
        {
        if (fscanf (fpnum,"%lf %lf", &za[i], &zb[i]) != 2)
	    {
	    Fprintf (stderr,
			"zero_pole: Not %d zeros in %s\n", nzero, num_file);
	    exit (1);
	    }
	}

    if (debug_level)
	{
	for (i=0; i<nzero; i++)
	    Fprintf (stderr,"zero_pole: z%d = %g + %g i\n", i, za[i], zb[i]);
	}

 /*
  * Get denominator coefficients, if appropriate
 */
    if (argc - optind > 1)
      {
	den_file = argv [optind++];
	dflag = YES;
	if (strcmp (den_file, "-") == 0) 
	    {
	    den_file = "<stdin>";
	    fpden = stdin;
	    cflag = YES; /* can't use standard input for comment*/
	    }
	else if (strcmp (den_file, num_file) != 0)
	    {
	    TRYOPEN("zero_pole", den_file, "r", fpden);
	    }
	else
	    {
	    fpden = fpnum;
	    }

	if (debug_level) 
	    Fprintf (stderr,"zero_pole: Pole file is %s\n", den_file);

        if (fscanf (fpden,"%d", &npole) != 1)
	    {
	    Fprintf (stderr,
		"zero_pole: Number of poles not specified.\n");
	    exit (1);
	    }
	if (npole < 0)
	    {
	    Fprintf (stderr,"npole = %d is unacceptable - must be > zero.\n"
			, npole);
	    exit (1);
	    }
	if (debug_level)
            Fprintf (stderr,"npole = %d\n", npole);

/* Allocate space for the coefficients. */

        pa = (double*) calloc ((unsigned)npole, sizeof (double));
    spsassert(pa != NULL, 
		"Couldn't allocate memory for pole coefficients\n");
        pb = (double*) calloc ((unsigned)npole, sizeof (double));
    spsassert(pb != NULL, 
		"Couldn't allocate memory for pole coefficients\n");

/* read the coefficients. */

        for (i=0; i<npole; i++)
	    {
            if (fscanf (fpden,"%lf %lf", &pa[i], &pb[i]) != 2)
	        {
	        Fprintf (stderr,
		    "zero_pole: Not %d poles in %s\n", npole, den_file);
	        exit (1);
	        }
	    }
        if (debug_level)
	    {
	    for (i=0; i<npole; i++)
	        Fprintf (stderr, "zero_pole: p%d = %g + %g i\n", i, pa[i], pb[i]);
	    }

      }

 /*
  * Now get output FEAFILT file name
 */
	{
	filt_file = eopen("zero_pole", argv[optind], "w", FT_FEA, FEA_FILT,
			    (struct header **)NULL, &fpout);

	if (debug_level) Fprintf (stderr,
				"zero_pole: Output file is %s\n", filt_file);
	}


/* Get comments from standard input. */

    if (cflag == NO)
	{
	(void)printf ("zero_pole: Please enter comments.\n");
	(void)printf ("zero_pole: Must be less than %d characters and terminated with ^D.\n", MX_S);
	(void)fflush (stdout);
	i = 0;
	while ((int)(comment3[i] = getchar()) != EOF && i < MX_S)
	    {
	    i ++;
	    }

/* Make sure the comment ends in a \n and a \0. */

	if (comment3[i-1] != '\n')
	    {
	    comment3[i] = '\n';
	    i ++;
	    }
	comment3 [i] = '\0';
	cflag = YES;
	}

/* Create the output header and set key values. */

    oh = new_header (FT_FEA);
    oh->common.tag = YES;
    Strcpy (oh->common.prog, "zero_pole");
    Strcpy (oh->common.vers, Version);
    Strcpy (oh->common.progdate, Date);

    fdp = (filtdesparams *) calloc( (unsigned) 1, sizeof(filtdesparams) );
    spsassert( fdp != NULL, "zero_pole: can't allocate header info memory.");

    fdp->method = PZ_PLACE;
    fdp->define_pz = YES;
    fdp->filter_complex = NO;
    fdp->type = FILT_ARB;
    fdp->func_spec = NONE;


/* Fill in the comment field. */

    comment1 = (char*) calloc ((unsigned)(COM_S+strlen(num_file)), sizeof (char));
    spsassert(comment1 != NULL, "Couldn't allocate memory for comment\n");
    (void)sprintf (comment1,"Zeros came from the file %s.\n", num_file);
    if (strlen (comment1) > MAX_STRING)
	{
	comment1[MAX_STRING - 1] = '\0';
	Fprintf (stderr,"zero_pole: Warning, comment field was truncated.");
	}
    else
	{
        oh->variable.comment = comment1;
        if (dflag == YES)
	    {
            comment2 = (char*) calloc ((unsigned)(COM_S+strlen(den_file)), 
					sizeof (char));
	    spsassert(comment2 != NULL, 
		    "Couldn't allocate memory for pole comment\n");
            (void)sprintf (comment2,
		    "Poles came from the file %s.\n", den_file);
	    add_comment (oh, comment2);
	    }
	}

    add_comment(oh, comment3);

/* Allocate the data record and fill in the values. */


    maxnum = 2*nzero + 1;
    maxdenom = 2*npole + 1;

    zptr = (float *) calloc( (unsigned) maxnum, sizeof(float));
    spsassert( zptr != NULL, "can't allocate memory for numerator poly.");

    if ( npole > 0 ) {
	pptr = (float *) calloc( (unsigned) maxdenom, sizeof(float));
	spsassert ( pptr != NULL, "can't allocate memory for denominator poly.");
      }

    init_feafilt_hd( oh, (long) maxnum, (long) maxdenom, fdp);

    write_header (oh, fpout);

    frec = allo_feafilt_rec (oh);

    (void)pz_to_co (nzero, za, zb, &n_dim, zptr);
    *frec->num_size = (long) n_dim;

    if (npole > 0)
	{
       (void)pz_to_co (npole, pa, pb, &d_dim, pptr);
       *frec->denom_size = (long) d_dim;
	/* Check stability. */
	for (i=0; i < npole; i++)
	    {
	    if (hypot(pa[i], pb[i]) >= 1.0)
		{
		Fprintf (stderr,"zero_pole:Warning, poles not inside the unit circle.\n");
		Fprintf (stderr,"p%d = %g + %g i\n", i, pa[i], pb[i]);
		}
	    }
	}
    else
      *frec->denom_size = (long) 0;

/* Normalize the gain if desired. */

    if (gain_flag)
	{
	rpart = ipart = 0.0;
	for (i=0; i < n_dim; i++)
	    {
	    rpart += zptr[i]*cos (2*PI*f*i);
	    ipart += zptr[i]*sin (2*PI*f*i);
	    }
	gain = hypot (rpart, ipart);
	if (d_dim > 0)
	    {
	    rpart = ipart = 0.0;
	    for (i=0; i < d_dim; i++)
	        {
	        rpart += pptr[i]*cos (2*PI*f*i);
	        ipart += pptr[i]*sin (2*PI*f*i);
	        }
	    if (rpart != 0.0 || ipart != 0.0)
	        gain /= hypot (rpart, ipart);
	    else
		gain = -1;
	    }
	if (debug_level)
	    {
	    Fprintf (stderr,"zero_pole: gain = %g (before normalization)\n", gain);
	    }
	if (gain >= 0.0 && gain < SMALL)
	    {
	    Fprintf (stderr,"zero_pole: gain at f=%lf is zero. Gain will not be changed. \n", f);
	    }
	else if (gain == -1)
	    {
	    Fprintf (stderr,"zero_pole: gain not adjusted.\n");
	    }
	else
	    {
	    for (i=0; i < n_dim; i++)
	        {
	        zptr[i] /= gain;
	        }
	    }
	}
    /* Write the data to the output file. */

    *frec->zero_dim = nzero;
    *frec->pole_dim = npole;

    for (i=0; i<n_dim; i++)
      frec->re_num_coeff[i] = (double) zptr[i];

    for (i=0; i<d_dim; i++)
      frec->re_denom_coeff[i] = (double) pptr[i];

    for (i=0; i<nzero; i++) {
	frec->zeros[i].real = za[i];
	frec->zeros[i].imag = zb[i];
      }

    for (i=0; i<npole; i++) {
	frec->poles[i].real = pa[i];
	frec->poles[i].imag = pb[i];
      }

    put_feafilt_rec (frec, oh, fpout);

    
    (void) fclose (fpout);
    exit(0);
    return(0); /* lint pleasing */
    }
